tests/transaction_hooks/tests.py

```diff
@@ -0,0 +1,223 @@
+                    with atomic():
+                        self.do(1)
+                    raise ForcedError()
+            except ForcedError:
+                pass
+            self.do(2)
+
+        self.assert_done([2])
+
+    def test_no_savepoints_atomic_merged_with_outer(self):
+        with atomic():
+            with atomic():
+                self.do(1)
+                try:
+                    with atomic(savepoint=False):
+                        raise ForcedError()
+                except ForcedError:
+                    pass
+
+        self.assert_done([])
+
+    def test_inner_savepoint_does_not_affect_outer(self):
+        with atomic():
+            with atomic():
+                self.do(1)
+                try:
+                    with atomic():
+                        raise ForcedError()
+                except ForcedError:
+                    pass
+
+        self.assert_done([1])
+
+    def test_runs_hooks_in_order_registered(self):
+        with atomic():
+            self.do(1)
+            with atomic():
+                self.do(2)
+            self.do(3)
+
+        self.assert_done([1, 2, 3])
+
+    def test_hooks_cleared_after_successful_commit(self):
+        with atomic():
+            self.do(1)
+        with atomic():
+            self.do(2)
+
+        self.assert_done([1, 2])  # not [1, 1, 2]
+
+    def test_hooks_cleared_after_rollback(self):
+        try:
+            with atomic():
+                self.do(1)
+                raise ForcedError()
+        except ForcedError:
+            pass
+
+        with atomic():
+            self.do(2)
+
+        self.assert_done([2])
+
+    @skipUnlessDBFeature('test_db_allows_multiple_connections')
+    def test_hooks_cleared_on_reconnect(self):
+        with atomic():
+            self.do(1)
+            connection.close()
+
+        connection.connect()
+
+        with atomic():
+            self.do(2)
+
+        self.assert_done([2])
+
+    def test_error_in_hook_doesnt_prevent_clearing_hooks(self):
+        try:
+            with atomic():
+                connection.on_commit(lambda: self.notify('error'))
+        except ForcedError:
+            pass
+
+        with atomic():
+            self.do(1)
+
+        self.assert_done([1])
+
+    def test_db_query_in_hook(self):
+        with atomic():
+            Thing.objects.create(num=1)
+            connection.on_commit(
+                lambda: [self.notify(t.num) for t in Thing.objects.all()])
+
+        self.assert_done([1])
+
+    # On databases that don't work with autocommit off (SQLite), Atomic doesn't
+    # ever call set_autocommit(True) after committing a transaction (it sets
+    # ``connection.autocommit = True`` directly instead), so we have to run
+    # hooks immediately on commit instead of waiting for autocommit to be
+    # restored. If a hook tries to create an internal transaction of its own,
+    # this fails with an error. Need a better solution here; could be fixed if
```

Comment:
Oh, there is a fourth option: turn the `autocommit` flag on the SQLite backend into a property, so that we can hook into `connection.autocommit = True` to run the transaction hooks.